frame/election-provider-multi-phase/test-staking-e2e/src/mock.rs

```diff
@@ -0,0 +1,745 @@
+	type RewardRemainder = ();
+	type RuntimeEvent = RuntimeEvent;
+	type Slash = (); // burn slashes
+	type Reward = (); // rewards are minted from the void
+	type SessionsPerEra = SessionsPerEra;
+	type BondingDuration = BondingDuration;
+	type SlashDeferDuration = SlashDeferDuration;
+	type SlashCancelOrigin = EnsureRoot<AccountId>; // root can cancel slashes
+	type SessionInterface = Self;
+	type EraPayout = ();
+	type NextNewSession = Session;
+	type MaxNominatorRewardedPerValidator = MaxNominatorRewardedPerValidator;
+	type OffendingValidatorsThreshold = OffendingValidatorsThreshold;
+	type ElectionProvider = ElectionProviderMultiPhase;
+	type GenesisElectionProvider = onchain::OnChainExecution<OnChainSeqPhragmen>;
+	type VoterList = BagsList;
+	type TargetList = pallet_staking::UseValidatorsMap<Self>;
+	type MaxUnlockingChunks = ConstU32<32>;
+	type HistoryDepth = HistoryDepth;
+	type OnStakerSlash = ();
+	type WeightInfo = pallet_staking::weights::SubstrateWeight<Runtime>;
+	type BenchmarkingConfig = StakingBenchmarkingConfig;
+}
+
+impl<LocalCall> frame_system::offchain::SendTransactionTypes<LocalCall> for Runtime
+where
+	RuntimeCall: From<LocalCall>,
+{
+	type OverarchingCall = RuntimeCall;
+	type Extrinsic = Extrinsic;
+}
+
+pub struct OnChainSeqPhragmen;
+
+parameter_types! {
+	pub VotersBound: u32 = 600;
+	pub TargetsBound: u32 = 400;
+}
+
+impl onchain::Config for OnChainSeqPhragmen {
+	type System = Runtime;
+	type Solver = SequentialPhragmen<
+		AccountId,
+		pallet_election_provider_multi_phase::SolutionAccuracyOf<Runtime>,
+	>;
+	type DataProvider = Staking;
+	type WeightInfo = ();
+	type MaxWinners = MaxWinners;
+	type VotersBound = VotersBound;
+	type TargetsBound = TargetsBound;
+}
+
+pub struct StakingBenchmarkingConfig;
+impl pallet_staking::BenchmarkingConfig for StakingBenchmarkingConfig {
+	type MaxNominators = traits::ConstU32<1000>;
+	type MaxValidators = traits::ConstU32<1000>;
+}
+
+pub struct ElectionProviderBenchmarkConfig;
+
+impl pallet_election_provider_multi_phase::BenchmarkingConfig for ElectionProviderBenchmarkConfig {
+	const VOTERS: [u32; 2] = [1000, 2000];
+	const TARGETS: [u32; 2] = [500, 1000];
+	const ACTIVE_VOTERS: [u32; 2] = [500, 800];
+	const DESIRED_TARGETS: [u32; 2] = [200, 400];
+	const SNAPSHOT_MAXIMUM_VOTERS: u32 = 1000;
+	const MINER_MAXIMUM_VOTERS: u32 = 1000;
+	const MAXIMUM_TARGETS: u32 = 300;
+}
+
+pub struct OtherSessionHandler;
+impl traits::OneSessionHandler<AccountId> for OtherSessionHandler {
+	type Key = testing::UintAuthorityId;
+
+	fn on_genesis_session<'a, I: 'a>(_: I)
+	where
+		I: Iterator<Item = (&'a AccountId, Self::Key)>,
+		AccountId: 'a,
+	{
+	}
+
+	fn on_new_session<'a, I: 'a>(_: bool, _: I, _: I)
+	where
+		I: Iterator<Item = (&'a AccountId, Self::Key)>,
+		AccountId: 'a,
+	{
+	}
+
+	fn on_disabled(_validator_index: u32) {}
+}
+
+impl sp_runtime::BoundToRuntimeAppPublic for OtherSessionHandler {
+	type Public = testing::UintAuthorityId;
+}
+
+pub struct ExtBuilder {
```

Comment:
how about a nested `ExtBuilder` like: 

```

struct ExtBuilder {
    // things that are beyond one pallet.
    initialize_first_session: bool
    // things that are related to staking.
    staking_builder: StakingBuilder,
    // things that are related to epm.
    epm_builder: EPMBuilder,
}


```

Although this would add a lot of verbosity, perhaps you can examine yourself if it is worth is.

as a layman's alternative, you can prefix everything with `staking_` or `epm_`.

Reply:
After experimenting a bit, I think that splitting the externalities builder per pallet makes sense in this context (may be an overkill for mocks dedicated to one pallet only). The current PR contains the ext builder split.